src/timer_thread.c

```diff
@@ -0,0 +1,452 @@
+#endif
+#ifdef SIGXFSZ
+               {"XFSZ", SIGXFSZ},
+#endif
+#ifdef SIGVTALRM
+               {"VTALRM", SIGVTALRM},
+#endif
+#ifdef SIGPROF
+               {"PROF", SIGPROF},
+#endif
+#ifdef SIGWINCH
+               {"WINCH", SIGWINCH},
+#endif
+#ifdef SIGUSR1
+               {"USR1", SIGUSR1},
+#endif
+#ifdef SIGUSR2
+               {"USR2", SIGUSR2},
+#endif
+#ifdef SIGLOST
+               {"LOST", SIGLOST},
+#endif
+#ifdef SIGMSG
+               {"MSG", SIGMSG},
+#endif
+#ifdef SIGPWR
+               {"PWR", SIGPWR},
+#endif
+#ifdef SIGPOLL
+               {"POLL", SIGPOLL},
+#endif
+#ifdef SIGDANGER
+               {"DANGER", SIGDANGER},
+#endif
+#ifdef SIGMIGRATE
+               {"MIGRATE", SIGMIGRATE},
+#endif
+#ifdef SIGPRE
+               {"PRE", SIGPRE},
+#endif
+#ifdef SIGGRANT
+               {"GRANT", SIGGRANT},
+#endif
+#ifdef SIGRETRACT
+               {"RETRACT", SIGRETRACT},
+#endif
+#ifdef SIGSOUND
+               {"SOUND", SIGSOUND},
+#endif
+#ifdef SIGINFO
+               {"INFO", SIGINFO},
+#endif
+               {NULL, 0}};
+
+static int signm2signo(const char *nm)
+{
+  const struct signals *sigs;
+
+  for (sigs = siglist; sigs->signm; sigs++) {
+    if (strcmp(sigs->signm, nm) == 0)
+      return sigs->signo;
+  }
+
+  /* Handle RT Signal#0 as special for strtol's err spec */
+  if (strcmp("RT0", nm) == 0)
+    return SIGRTMIN;
+
+  if (strncmp("RT", nm, 2) == 0) {
+    int ret = (int)strtol(nm + 2, NULL, 0);
+    if (!ret || (SIGRTMIN + ret > SIGRTMAX))
+      return 0;
+    return SIGRTMIN + ret;
+  }
+  return 0;
+}
+
+static int mrb_to_signo(mrb_state *mrb, mrb_value vsig)
+{
+  int sig = -1;
+  const char *s;
+
+  switch (mrb_type(vsig)) {
+  case MRB_TT_FIXNUM:
+    sig = mrb_fixnum(vsig);
+    if (sig < 0 || sig >= SIGRTMAX) {
+      mrb_raisef(mrb, E_ARGUMENT_ERROR, "invalid signal number (%S)", vsig);
+    }
+    break;
+  case MRB_TT_SYMBOL:
+    s = mrb_sym2name(mrb, mrb_symbol(vsig));
+    if (!s)
+      mrb_raise(mrb, E_ARGUMENT_ERROR, "bad signal");
+    goto str_signal;
+  default:
+    vsig = mrb_string_type(mrb, vsig);
+    s = RSTRING_PTR(vsig);
+
+  str_signal:
+    if (memcmp("SIG", s, 3) == 0)
+      s += 3;
+    sig = signm2signo(s);
+    if (sig == 0 && strcmp(s, "EXIT") != 0)
+      mrb_raise(mrb, E_ARGUMENT_ERROR, "unsupported signal");
+    break;
+  }
+  return sig;
+}
+
+typedef struct {
+  timer_t *timer_ptr;
+  int timer_signo;
+} mrb_timer_posix_data;
+
+static void mrb_timer_posix_free(mrb_state *mrb, void *p)
+{
+  mrb_timer_posix_data *data = (mrb_timer_posix_data *)p;
+  timer_delete(*data->timer_ptr);
+  mrb_free(mrb, data->timer_ptr);
+  mrb_free(mrb, data);
+}
+
+static const struct mrb_data_type mrb_timer_posix_data_type = {"mrb_timer_posix_data", mrb_timer_posix_free};
+
+static mrb_value mrb_rtsignal_get(mrb_state *mrb, mrb_value self)
+{
+  mrb_int idx;
+  if (mrb_get_args(mrb, "i", &idx) == -1) {
+    mrb_raise(mrb, E_RUNTIME_ERROR, "Cannot get arguments");
+  }
```

Comment:
It is better to unify it to either `if (hoge) fuga;` or `if (hoge) { fuga; }` notation.